### ISSUE TITLE
Break up Prelude

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1131]
From IRC discussion.

- Restoring the explicit `Prelude` import is desirable to avoid giving it special significance in being able to claim names and operators.
- Following on from that, we may as well break the `Prelude` up, since if it's not magic anymore there's no restriction on it being able to import things into itself (and then re-export them), and it gives us more freedom to define better class hierarchies and the like without ending up with a gigantic `Prelude` module.
- Having some kind of `Prelude` module is still desirable, as importing it should provide a generally useful environment in which to get things done without having to import tons of different modules.
- The `--no-prelude` flag will no longer be necessary once this is done.

Unfortunately we can't do re-exports properly yet, so my suggestion is maybe for now we just make the `Prelude` import explicit, but keep most or all of its current contents, and then later pull things out of it when re-exporting is available.

**Comment by puffnfresh** [comment]
:+1:

**Comment by michaelficarra** [comment]
:+1:

**Comment by vendethiel** [comment]
@michaelficarra that's the 4th bullet :)

It sounds like the best option.

**Comment by michaelficarra** [comment]
@vendethiel http://ircbrowse.net/browse/purescript?id=107613&timestamp=1432766488#t1432766488

**Comment by paf31** [comment]
:+1: Since this doesn't pollute the default namespace any more, we can make the `BoolLike` operations monomorphic in Prelude now, yes?

**Comment by jdegoes** [comment]
:+1:

**Comment by freebroccolo** [comment]
Sounds good :+1:

**Comment by hdgarrood** [comment]
Is the `--no-prelude` flag already unnecessary (in master at least) since the Prelude is no longer embedded in the compiler?

**Comment by paf31** [comment]
`--no-prelude` just turns the import off.

**Comment by paf31** [comment]
I'll move this to the new `Discussion` milestone.

**Comment by hdgarrood** [comment]
Time to close this, I think? All the point have been implemented now, apart from breaking up the prelude (which looks like it is about to happen, and besides, is already tracked on the prelude repo).
